cpp/memilio/io/history.h

```diff
@@ -120,50 +102,59 @@ class History
         return m_data;
     }
 
-    template <class Logger>
-    const std::vector<typename Logger::Type>& get_log()
-    {
-        return std::get<details::index_templ_pack<Logger, Loggers...>()>(m_data);
-    }
-
 private:
     typename WriteWrapper::Data m_data;
+    std::tuple<Loggers...> m_loggers;
 
-    bool logged = false;
-
-    template <class T, class logger, class... loggers>
-    std::enable_if_t<std::is_base_of<LogOnceStart, logger>::value> log_impl(const T& t)
+    /**
+     * @brief Checks if the given logger should log. If so, adds a record of the log to m_data.
+     * @param[in] t The argument given to History::log. Passed to Logger::should_log and Logger::log.
+     * @param[in] logger A Logger instance.
+     * @tparam Logger A logger from the list Loggers.
+     */
+    template <class T, class Logger>
+    void log_impl(const T& t, Logger& logger)
     {
-        if (!logged) {
-            WriteWrapper::template log_this<logger>(logger::log(t), m_data);
+        if (logger.should_log(t)) {
+            WriteWrapper::template add_record<Logger>(logger.log(t), m_data);
```

Comment:
We added a short comment to the Writer. The main documentation for these requirements is in the History class and io/Readme.

> Also, can we somehow extract the std::get<mio::index_of_type_v<Logger, Loggers...>>(data) from the writer

Yes, technically, but I don't think we should. The problem is, that the Data tuple only stores the types Logger::Type, which are in general not unique. So we cannot use `std::get<Logger::Type>` or anything similar, we have to look up the `Logger`s index in `Loggers...`. We can change around how we do it, but we need either those two templates, or (somehow) the index of Logger, so that we can call `std::get<size_t>`. We can for example store `Loggers...` in a TypeList, or add a member function to the Writer (or History, which would have its own Problems) to replace the get call, but that would only move what we already do around a bit, which makes the get<index<...>> stuff even more obscure.

If we require that all Data types have to be tuples of length sizeof...(Loggers), then we can add History::get_log<Logger> back (or something similar for the writer to use). But that is a big restriction, it would bar us from e.g. using .